RecentActivity/src/org/sleuthkit/autopsy/recentactivity/BinaryCookieReader.java

```diff
@@ -0,0 +1,429 @@
+
+    private final int[] pageSizeArray;
+    private final File cookieFile;
+
+    private final Logger logger = Logger.getLogger(this.getClass().getName());
+
+    /**
+     * SafariCookieReader wraps all the knowledge of how to read the mac
+     * .binarycookie files into one class.
+     *
+     */
+    private BinaryCookieReader(File file, int[] sizeArray) {
+        cookieFile = file;
+        pageSizeArray = sizeArray;
+    }
+
+    /**
+     * initalizeReader opens the given file, reads the header and checks that
+     * the file is a binarycookie file. This function does not keep the file
+     * open.
+     *
+     * @param file binarycookie file
+     * @return An instance of the reader
+     * @throws FileNotFoundException
+     * @throws IOException
+     */
+    public static BinaryCookieReader initalizeReader(File file) throws FileNotFoundException, IOException {
+        BinaryCookieReader reader = null;
+        try (DataInputStream dataStream = new DataInputStream(new FileInputStream(file))) {
+
+            byte[] magic = new byte[MAGIC_SIZE];
+            if (dataStream.read(magic) != MAGIC_SIZE) {
+                throw new IOException("Failed to read header, invalid file size" + file.getName()); //NON-NLS
+            }
+
+            if (!(new String(magic)).equals(COOKIE_MAGIC)) {
+                throw new IOException(file.getName() + " is not a cookie file"); //NON-NLS
+            }
+
+            int[] sizeArray = null;
+            int pageCount = dataStream.readInt();
+            if (pageCount != 0) {
+                sizeArray = new int[pageCount];
+
+                for (int cnt = 0; cnt < pageCount; cnt++) {
+                    sizeArray[cnt] = dataStream.readInt();
+                }
+            }
+
+            reader = new BinaryCookieReader(file, sizeArray);
+        }
+
+        return reader;
+    }
+
+    /**
+     * Creates and returns a instance of CookiePageIterator
+     *
+     * @return CookiePageIterator
+     */
+    @Override
+    public Iterator<Cookie> iterator() {
+        return new CookiePageIterator();
+    }
+
+    /**
+     * The cookiePageIterator iterates the binarycookie file by page.
+     */
+    private class CookiePageIterator implements Iterator<Cookie> {
+
+        int pageIndex = 0;
+        CookiePage currentPage = null;
+        Iterator<Cookie> currentIterator = null;
+        DataInputStream dataStream = null;
+
+        /**
+         * The cookiePageIterator iterates the binarycookie file by page.
+         */
+        CookiePageIterator() {
+            try {
+                dataStream = new DataInputStream(new FileInputStream(cookieFile));
+                // skip to the first page
+                dataStream.skipBytes((2 * SIZEOF_INT_BYTES) + (pageSizeArray.length * SIZEOF_INT_BYTES));
+            } catch (IOException ex) {
+                logger.log(Level.WARNING, "Error occured creating DataInputStream", new Object[]{cookieFile.getName(), ex}); //NON-NLS
```

Comment:
I missed the place holder which I will fix, but there is no four argument version of log, at least not in Java 7.  How would you include the exception and an argument in the message?  

https://docs.oracle.com/javase/7/docs/api/java/util/logging/Logger.html

Reply:
What I would do is have a message along the lines of "An error occurred creating an input stream for X (objId=N)" since putting the exception message in the log message will be a bit redundant since the stack trace of the exception, which will show the exception message. The information you are adding is unambiguous identification of the file that experienced the error (note that the bit about logging the object id of the file assumes it is a file in the case DB).

Another technique you can use is to make the message using String.format, e.g., 
String.format("An error occurred creating an input stream for %s (objId=%d): %s", fileName, objId, ex.getMessage())

Reply:
String.format is what I needed! Thanks!